File: spring-data-requery-test-java/src/main/java/org/springframework/data/requery/domain/sample/AbstractParent.java
package org.springframework.data.requery.domain.sample;

import io.requery.*;
import lombok.Getter;
import org.springframework.data.requery.domain.AbstractPersistable;

import java.util.Set;

/**
 * AbstractParent
 *
 * @author debop@coupang.com
 * @since 18. 6. 25
 */
@Getter
@Entity
public abstract class AbstractParent extends AbstractPersistable<Long> {

    @Key
    @Generated
    protected Long id;

    @JunctionTable
    @ManyToMany
    protected Set<AbstractChild> children;

    protected String name;

    public AbstractParent addChild(AbstractChild child) {
        this.getChildren().add(child);
        return this;
    }

    private static final long serialVersionUID = 7310565121245237400L;
}

File: src/test/kotlin/com/coupang/springframework/data/requery/domain/sample/AbstractParent.kt
package com.coupang.springframework.data.requery.domain.sample

import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*

/**
 * com.coupang.springframework.data.requery.domain.sample.AbstractParent
 * @author debop
 * @since 18. 5. 30
 */
@Entity
abstract class AbstractParent: AbstractPersistable<Long>() {

    @get:Key
    @get:Generated
    abstract override val id: Long?

    abstract var name: String?

    @get:ManyToMany
    @get:JunctionTable
    abstract val children: MutableSet<AbstractChild>

    fun addChild(child: AbstractChild): AbstractParent {
        this.children.add(child)

        if(!child.parents.contains(this)) {
            child.addParent(this)
        }
        return this
    }
}